addons/stock/models/product.py

```diff
@@ -761,6 +761,19 @@ def _onchange_type(self):
         return res
 
     def write(self, vals):
+        if 'company_id' in vals and vals['company_id']:
+            for product in self:
+                if product.company_id.id == vals['company_id']:
+                    continue
+            # Forbid changing the product's company when quants exist in another company.
+            existing_quants_count = self.env['stock.quant'].sudo().search([
+                ('product_id', 'in', product.product_variant_ids.ids),
+                ('company_id', '!=', vals['company_id']),
+                ('quantity', '!=', 0),
+            ], order=None, limit=1)
+            if existing_quants_count:
+                raise UserError(_("This product's company cannot be changed as long as there are quantities of it belonging to another company."))
```

Comment:
Now that I look at this again, I think it's missing an indent to avoid doing the search if none of the products are having their company changed:

```suggestion
                if product.company_id.id == vals['company_id']:
                    continue
                # Forbid changing the product's company when quants exist in another company.
                existing_quants_count = self.env['stock.quant'].sudo().search([
                    ('product_id', 'in', product.product_variant_ids.ids),
                    ('company_id', '!=', vals['company_id']),
                    ('quantity', '!=', 0),
                ], order=None, limit=1)
                if existing_quants_count:
                    raise UserError(_("This product's company cannot be changed as long as there are quantities of it belonging to another company."))
```

Also :+1: about adding the `order=None` just in case. I expect someone, someday, will change the company of a product when they have a large number of quants and will never know that we pre-optimized for it :angel:

Reply:
This missing indent was quite a bloody mistake... Thank you to notified it :sweat_smile:

Reply:
@ticodoo I was thinking... Maybe it's even better to do only one search:
```python
if 'company_id' in vals and vals['company_id']:
    changing_company_products = self.env['product.template']
    for product in self:
        if product.company_id.id != vals['company_id']:
            changing_company_products += product
    if changing_company_products:
        # Forbid changing the product's company when quants exist in another company.
        existing_quants_count = self.env['stock.quant'].sudo().search([
            ('product_id', 'in', changing_company_products.product_variant_ids.ids),
            ('company_id', '!=', vals['company_id']),
            ('quantity', '!=', 0),
        ], order=None, limit=1)
        if existing_quants_count:
            raise UserError(_("This product's company cannot be changed as long as there are quantities of it belonging to another company."))
```
Because if we write on multiple products, we don't want to write if it exists one quant for one of these products. Doing a single search for all products at once shouldn't change the final result, but it's avoid doing multiple searches :smile: